Clean build output for regular builds

epeee opened the pull request:
a very first draft for #664 in order to get early feedback.

epeee commented on subprojects/shipkit/src/main/groovy/org/shipkit/gradle/release/ReleaseNeededTask.java:
```diff
@@ -102,21 +101,6 @@ public void setSkipComparePublications(boolean skipComparePublications) {
-    /**
-     * If the exception should be thrown if the release is not needed.
-     */
-    public boolean isExplosive() {
```
since this one is public api, do we have to keep it and mark it deprecated? or can we remove it an create a 2.1.0?

mockitoguy replied:
We keep it and deprecate it. Can you create a static utility method like "deprecated(String message)" so that when we ship a new major version, we can wrap up all deprecations?

mockitoguy commented on subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/CiReleasePlugin.java:
```diff
@@ -51,7 +54,11 @@ public void apply(final Project project) {
-                        "Checking if release is needed", asList("./gradlew", ReleaseNeededPlugin.ASSERT_RELEASE_NEEDED_TASK), ExecCommandFactory.stopExecution()));
+                    "Checking if release is needed", asList("./gradlew", ReleaseNeededPlugin.RELEASE_NEEDED), execResult -> {
+                        if (new File(project.getBuildDir(), ReleaseNeeded.RELEASE_NEEDED_FILENAME).exists()) {
+                            throw new StopExecutionException();
```
This is a really cunning idea. Nice!

mockitoguy commented on subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/tasks/ReleaseNeeded.java:
```diff
@@ -28,11 +33,24 @@ boolean releaseNeeded(ReleaseNeededTask task, EnvVariables envVariables) {
-            LOG.lifecycle(message);
+        File releaseNeededFile = new File(task.getProject().getBuildDir(), RELEASE_NEEDED_FILENAME);
+
+        if (releaseNeededFile.exists()) {
```
Great! We need to delete the file.

Can you make it the first thing that happens in this task?

mockitoguy commented on subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/tasks/ReleaseNeeded.java:
```diff
@@ -28,11 +33,24 @@ boolean releaseNeeded(ReleaseNeededTask task, EnvVariables envVariables) {
+        }
+
+        if (releaseNeeded) {
+            try {
```
Use IOUtil.writeFile() and save 8 lines of code :)

epeee replied:
:+1: , thx for this hint!

mockitoguy commented on subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/tasks/ReleaseNeeded.java:
```diff
@@ -28,11 +33,24 @@ boolean releaseNeeded(ReleaseNeededTask task, EnvVariables envVariables) {
         String message = releaseNeed.explanation;
 
-        if (!releaseNeeded && task.isExplosive()) {
-            throw new GradleException(message);
```
For backwards compatibility, please keep the behavior.

mockitoguy commented on subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/ReleaseNeededPlugin.java:
```diff
@@ -34,32 +34,21 @@
-        //Task that throws an exception when release is not needed is very useful for CI workflows
-        //Travis CI job will stop executing further commands if assertReleaseNeeded fails.
-        //See the example projects how we have set up the 'assertReleaseNeeded' task in CI pipeline.
-        releaseNeededTask(project, ASSERT_RELEASE_NEEDED_TASK, conf)
```
For backwards compatibility, please keep the task. I suggest to emit deprecated message when the task is executed, including information how to fix the deprecation.

epeee commented:
@mockitoguy thx for your feedback.
The only part which I was not able to cover is the templates (initShipkit task) part. Maybe I have overlooked something but I could not find any parts which would require a change. Did I miss something?

mockitoguy commented:
I'll complete the review today! Thanks!

epeee commented:
I just did some local testing using shipkit-example project:
```
./gradlew ciPerformRelease
  Building version '0.16.14' (value loaded from 'version.properties' file).
  [INCUBATING] upgrade-dependency plugin is incubating and may change in any version.
:ciPerformRelease
  Checking if release is needed:
    ./gradlew releaseNeeded
[releaseNeeded] Starting a Gradle Daemon, 1 busy and 1 stopped Daemons could not be reused, use --status for details
[releaseNeeded]   Building version '0.16.14' (value loaded from 'version.properties' file).
[releaseNeeded]   [INCUBATING] upgrade-dependency plugin is incubating and may change in any version.
[releaseNeeded] :identifyGitBranch
[releaseNeeded]   Executing:
[releaseNeeded]     git rev-parse --abbrev-ref HEAD
[releaseNeeded]   Identified current branch: release/ep
[releaseNeeded] :api:downloadPreviousReleaseArtifacts
[releaseNeeded]   Downloading remote artifact
[releaseNeeded]   - from https://bintray.com/shipkit/examples/download_file?file_path=org/mockito/shipkit-example/api/0.16.13/api-0.16.13-sources.jar
[releaseNeeded]   - and saving it to /home/ep/tmp/shipkit-example/subprojects/api/build/previous-release-artifacts/api-0.16.13-sources.jar
[releaseNeeded] :api:createDependencyInfoFile
[releaseNeeded] :api:sourcesJar UP-TO-DATE
[releaseNeeded] :api:comparePublications
[releaseNeeded] :api:comparePublications - about to compare publications
[releaseNeeded] :api:comparePublications - META-INF/dependency-info.md files equal: true
[releaseNeeded] :api:comparePublications - source jars equal: true
[releaseNeeded] :api:comparePublications - You can find detailed publication comparison results in file /home/ep/tmp/shipkit-example/subprojects/api/build/publications-comparison.txt.
[releaseNeeded] :impl:downloadPreviousReleaseArtifacts
[releaseNeeded]   Downloading remote artifact
[releaseNeeded]   - from https://bintray.com/shipkit/examples/download_file?file_path=org/mockito/shipkit-example/impl/0.16.13/impl-0.16.13-sources.jar
[releaseNeeded]   - and saving it to /home/ep/tmp/shipkit-example/subprojects/impl/build/previous-release-artifacts/impl-0.16.13-sources.jar
[releaseNeeded] :api:compileJava UP-TO-DATE
[releaseNeeded] :api:processResources NO-SOURCE
[releaseNeeded] :api:classes UP-TO-DATE
[releaseNeeded] :api:jar UP-TO-DATE
[releaseNeeded] :impl:createDependencyInfoFile
[releaseNeeded] :impl:sourcesJar UP-TO-DATE
[releaseNeeded] :impl:comparePublications
[releaseNeeded] :impl:comparePublications - about to compare publications
[releaseNeeded] :impl:comparePublications - META-INF/dependency-info.md files equal: true
[releaseNeeded] :impl:comparePublications - source jars equal: true
[releaseNeeded] :impl:comparePublications - You can find detailed publication comparison results in file /home/ep/tmp/shipkit-example/subprojects/impl/build/publications-comparison.txt.
[releaseNeeded] :releaseNeeded
[releaseNeeded]   Commit message to inspect for keywords '[ci skip-release]' and '[ci skip-compare-publications]': <unknown commit message>
[releaseNeeded]   Current branch 'release/ep' matches 'master|release/.+': true
[releaseNeeded] 
[releaseNeeded]   Compared 2 publication(s). No changes since previous release!
[releaseNeeded] 
[releaseNeeded] Release is considered _not_ needed when:
[releaseNeeded]  - 'SKIP_RELEASE' environment variable is present (currently: false)
[releaseNeeded]  - commit message contains '[ci skip-release]' (currently: false)
[releaseNeeded]  - we are building a "pull request" (currently: false)
[releaseNeeded] Release is needed when all above is false and: 
[releaseNeeded]  - we are building on a branch that matches regex 'master|release/.+' (currently: true)
[releaseNeeded]  - and one of the following criteria is true:
[releaseNeeded]    - there are changes in the binaries when compared to previous version (currently: false)
[releaseNeeded]    - commit message contains '[ci skip-compare-publications]' (currently: false)
[releaseNeeded]    - 'skipComparePublications' property on task releaseNeeded is true (currently: false)
[releaseNeeded] 
[releaseNeeded]  Skipping release because publications are identical.
[releaseNeeded] 
[releaseNeeded] BUILD SUCCESSFUL
[releaseNeeded] 
[releaseNeeded] Total time: 10.369 secs
  External process [releaseNeeded] completed.

BUILD SUCCESSFUL

Total time: 12.484 secs
```
